File: lib/rgfa/sequence.rb
require_relative "error"

#
# Extensions of the String class to handle nucleotidic sequences
#
module RGFA::Sequence

  # Computes the reverse complement of a nucleotidic sequence
  #
  # @return [String] reverse complement, without newlines and spaces
  # @return [String] "*" if string is "*"
  #
  # @param valid [Boolean] <i>(defaults to: +false+)</i>
  #   if true, anything non-sequence is complemented to itself
  # @param rna [Boolean] <i>(defaults to: +false+)</i>
  #   if true, any A and a is complemented into u and U
  #
  # @raise [RGFA::ValueError] if chars are found for which
  #   no Watson-Crick complement is defined (and not +valid+)
  #
  # @example
  #  "ACTG".rc  # => "CAGT"
  #  "acGT".rc  # => "ACgt"
  # @example Undefined sequence is represented by "*":
  #  "*".rc     # => "*"
  # @example Extended IUPAC Alphabet:
  #  "ARBN".rc  # => "NVYT"
  # @example Usage with RNA sequences:
  #  "ACG".rc(rna: true) # => "CGU"
  def rc(valid: false, rna: false)
    return self if self.placeholder?
    retval = each_char.map do |c|
      wcc = WCC.fetch(c, valid ? c : nil)
      if wcc.nil?
        raise RGFA::ValueError,
          "#{self}: no Watson-Crick complement for #{c}"
      end
      wcc
    end.reverse.join
    retval.tr!("tT","uU") if rna
    retval
  end

  # Watson-Crick Complements
  WCC = {"a"=>"t","t"=>"a","A"=>"T","T"=>"A",
         "c"=>"g","g"=>"c","C"=>"G","G"=>"C",
         "b"=>"v","B"=>"V","v"=>"b","V"=>"B",
         "h"=>"d","H"=>"D","d"=>"h","D"=>"H",
         "R"=>"Y","Y"=>"R","r"=>"y","y"=>"r",
         "K"=>"M","M"=>"K","k"=>"m","m"=>"k",
         "S"=>"S","s"=>"s","w"=>"w","W"=>"W",
         "n"=>"n","N"=>"N","u"=>"a","U"=>"A",
         "-"=>"-","."=>".","="=>"=",
         " "=>"","\n"=>""}

  # Parse a string as sequence.
  # @return [RGFA::Placeholder, self] returns self if the string content
  #   is other than "*", otherwise a RGFA::Placeholder object
  def to_sequence
    self.placeholder? ? RGFA::Placeholder.new : self
  end
end

class String
  include RGFA::Sequence
end

File: gfapy/sequence.py
"""
Methods for processing strings as nucleotidic sequences
"""
import gfapy

def rc(sequence, valid = False, rna = False):
  """Compute the reverse complement of a nucleotidic sequence.

  All characters in the IUPAC extended alphabet are supported
  (ACGTUBVHDRYKMSWN). The character ".-=", spaces and newlines
  are left as they are. The case of each character is preserved.

  Returns
    str : reverse complement, without newlines and spaces;
         	"*" if string is "*"

  Parameters:
    sequence (str) : the sequence to reverse-complement
    valid (bool) : if True, the reverse complement of any invalid character
      is the character itself
    rna (bool) : if True, t/T are substituted by u/U in the output

  Raises:
    gfapy.error.ValueError : if valid is False and an invalid character
      (not in the IUPAC extended alphabet for nucleotide sequences, .-=,
      spaces or newline) is found
  """
  if gfapy.is_placeholder(sequence): return sequence
  def fun(c):
    wcc = WCC.get(c, c if valid else None)
    if not wcc:
      raise gfapy.ValueError("{}: no Watson-Crick complement for {}".format(sequence, c))
    return wcc
  retval = "".join(reversed([ fun(c) for c in sequence ]))
  if rna:
    retval = retval.translate(str.maketrans("tT", "uU"))
  return retval

WCC = {"a":"t","t":"a","A":"T","T":"A",
       "c":"g","g":"c","C":"G","G":"C",
       "b":"v","B":"V","v":"b","V":"B",
       "h":"d","H":"D","d":"h","D":"H",
       "R":"Y","Y":"R","r":"y","y":"r",
       "K":"M","M":"K","k":"m","m":"k",
       "S":"S","s":"s","w":"w","W":"W",
       "n":"n","N":"N","u":"a","U":"A",
       "-":"-",".":".","=":"=",
       " ":"","\n":""}
"""Watson-Crick Complements"""

def Sequence(string):
  """Parses the content of a sequence field.

  Parameters:
    string (str) : content of a sequence field

  Returns:
    str, gfapy.Placeholder : if the string is the placeholder
      symbol ``*`` then a placeholder, otherwise the string
      itself
  """
  return gfapy.Placeholder() if (string == "*") else string
